aws-s3-transfer-manager/src/runtime/token_bucket.rs

```diff
@@ -0,0 +1,245 @@
+
+    /// Acquire a token for the given permit type. Tokens are returned to the bucket when the
+    /// [OwnedToken] is dropped.
+    pub(crate) fn acquire(&self, ptype: PermitType) -> AcquireTokenFuture {
+        AcquireTokenFuture::new(PollSemaphore::new(self.semaphore.clone()), self.cost(ptype))
+    }
+
+    pub(crate) fn try_acquire(
+        &self,
+        ptype: PermitType,
+    ) -> Result<Option<OwnedToken>, error::Error> {
+        let cost = self.cost(ptype);
+        match self.semaphore.clone().try_acquire_many_owned(cost) {
+            Ok(permit) => Ok(Some(OwnedToken::new(permit))),
+            Err(TryAcquireError::NoPermits) => Ok(None),
+            Err(err @ TryAcquireError::Closed) => {
+                Err(error::Error::new(error::ErrorKind::RuntimeError, err))
+            }
+        }
+    }
+}
+
+pin_project! {
+    #[derive(Debug, Clone)]
+    pub(crate) struct AcquireTokenFuture {
+        sem: PollSemaphore,
+        tokens: u32
+    }
+}
+
+impl AcquireTokenFuture {
+    fn new(sem: PollSemaphore, tokens: u32) -> Self {
+        Self { sem, tokens }
+    }
+}
+
+impl Future for AcquireTokenFuture {
+    type Output = Result<OwnedToken, error::Error>;
+
+    fn poll(
+        self: std::pin::Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Self::Output> {
+        let this = self.project();
+        match this.sem.poll_acquire_many(cx, *this.tokens) {
+            Poll::Ready(Some(permit)) => Poll::Ready(Ok(OwnedToken::new(permit))),
+            Poll::Ready(None) => Poll::Ready(Err(error::Error::new(
+                error::ErrorKind::RuntimeError,
+                "semaphore closed",
+            ))),
+            Poll::Pending => Poll::Pending,
+        }
+    }
+}
+
+/// An owned permit from the scheduler to perform some unit of work.
+#[must_use]
+#[clippy::has_significant_drop]
+#[derive(Debug)]
+pub(crate) struct OwnedToken {
+    _inner: OwnedSemaphorePermit,
+}
+
+impl OwnedToken {
+    fn new(permit: OwnedSemaphorePermit) -> Self {
+        OwnedToken { _inner: permit }
+    }
+}
+
+/// Get the token bucket size to use for a given target throughput
+fn token_bucket_size(throughput: Throughput) -> usize {
+    let megabit_per_sec = throughput.as_unit_per_sec(ByteUnit::Megabit).max(1.0) as usize;
+    cmp::max(MIN_TOKENS, megabit_per_sec)
+}
+
+/// Tokens for payload size
+fn tokens_for_payload(payload_size: u64) -> usize {
+    let estimated_mbps = estimated_throughput(
+        payload_size,
+        S3_P50_REQUEST_LATENCY_MS,
+        Throughput::new_bytes_per_sec(S3_MAX_PER_REQUEST_THROUGHPUT_MB_PER_SEC * 1000 * 1000),
+    )
+    .as_unit_per_sec(ByteUnit::Megabit)
+    .round()
+    .max(1.0) as usize;
+
+    cmp::max(estimated_mbps, MIN_PAYLOAD_COST_TOKENS)
+}
+
+/// Estimate the throughput of a given request payload based on S3 latencies
+/// and max per/connection estimates.
+fn estimated_throughput(
+    payload_size: u64,
+    estimated_p50_latency_ms: usize,
```

Comment:
trivial: it's so weird that Semaphore::new() takes usize, but acquire_many() takes u32. It leads to a crazy mix of u32, usize, u64, etc based on how a variable or const will eventually be used (i.e. passing time here as usize)?

Can we try and reduce the number of types in use? Like stick with u32 or u64 whenever possible, and just clamp/cast at the last moment where necessary?

Reply:
Ughh yeah I don't understand why acquire takes a u32, I'm sure there's a reason but it's annoying. I'll look at what makes sense.